dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/ValidationUtils.java

```diff
@@ -99,24 +100,48 @@ public static List<Note> validateNotes(
     return notes;
   }
 
-  public static List<MetadataIdentifier> validateMandatoryDataValue(
-      ProgramStage programStage, Event event, List<MetadataIdentifier> mandatoryDataElements) {
-    List<MetadataIdentifier> notPresentMandatoryDataElements = Lists.newArrayList();
-
+  public static List<MetadataIdentifier> validateDeletionMandatoryDataValue(
+      Event event, ProgramStage programStage, List<MetadataIdentifier> mandatoryDataElements) {
     if (!needsToValidateDataValues(event, programStage)) {
-      return notPresentMandatoryDataElements;
+      return List.of();
+    }
+    Set<MetadataIdentifier> eventDataElements =
+        event.getDataValues().stream()
+            .filter(dv -> dv.getValue() == null)
+            .map(DataValue::getDataElement)
+            .collect(Collectors.toSet());
+
+    return mandatoryDataElements.stream().filter(eventDataElements::contains).toList();
+  }
+
+  public static List<MetadataIdentifier> validateMandatoryDataValue(
+      TrackerBundle bundle,
+      Event event,
+      ProgramStage programStage,
+      List<MetadataIdentifier> mandatoryDataElements) {
+    if (!areDataValuesBeingCreated(bundle, event, programStage)) {
+      return List.of();
     }
 
     Set<MetadataIdentifier> eventDataElements =
         event.getDataValues().stream().map(DataValue::getDataElement).collect(Collectors.toSet());
 
-    for (MetadataIdentifier mandatoryDataElement : mandatoryDataElements) {
-      if (!eventDataElements.contains(mandatoryDataElement)) {
-        notPresentMandatoryDataElements.add(mandatoryDataElement);
-      }
+    return mandatoryDataElements.stream().filter(de -> !eventDataElements.contains(de)).toList();
+  }
+
+  public static boolean areDataValuesBeingCreated(
+      TrackerBundle bundle, Event event, ProgramStage programStage) {
+    if (!needsToValidateDataValues(event, programStage)) {
+      return false;
+    }
+
+    if (bundle.getStrategy(event).isCreate()) {
+      return true;
     }
 
-    return notPresentMandatoryDataElements;
+    EventStatus savedStatus = bundle.getPreheat().getEvent(event.getUid()).getStatus();
+    return EventStatus.STATUSES_WITHOUT_DATA_VALUES.contains(savedStatus)
```

Comment:
do the event statuses `OVERDUE` and `SKIPPED` also need to be considered here? or `SCHEDULE` would be enough?

Reply:
Yes they do. They have the exact same behaviour as `SCHEDULE` events, so we need to check them to be sure we are in a consistent state